projects/steering/test/test_steering_digital_input.c

```diff
@@ -0,0 +1,115 @@
+#include "can.h"
+#include "can_msg_defs.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "interrupt_def.h"
+#include "ms_test_helpers.h"
+#include "soft_timer.h"
+#include "status.h"
+#include "steering_can.h"
+#include "steering_digital_input.h"
+#include "steering_events.h"
+#include "test_helpers.h"
+
+#define STEERING_CAN_DEVICE_ID 0x1
+
+typedef enum {
+  STEERING_CAN_EVENT_RX = 10,
+  STEERING_CAN_EVENT_TX,
+  STEERING_CAN_FAULT,
+} SteeringCanEvent;
+
+static CanStorage s_can_storage;
+
+static CanSettings can_settings = { .device_id = STEERING_CAN_DEVICE_ID,
+                                    .bitrate = CAN_HW_BITRATE_500KBPS,
+                                    .rx_event = STEERING_CAN_EVENT_RX,
+                                    .tx_event = STEERING_CAN_EVENT_TX,
+                                    .fault_event = STEERING_CAN_FAULT,
+                                    .tx = { GPIO_PORT_A, 12 },
+                                    .rx = { GPIO_PORT_A, 11 },
+                                    .loopback = true };
+
+static int count = 0;
+
+StatusCode prv_test_horn_rx_cb_handler(const CanMessage *msg, void *context,
+                                       CanAckStatus *ack_reply) {
+  TEST_ASSERT_EQUAL(SYSTEM_CAN_MESSAGE_HORN, msg->msg_id);
+  count++;
+  return STATUS_CODE_OK;
+}
+
+StatusCode prv_test_high_beam_rx_cb_handler(const CanMessage *msg, void *context,
+                                            CanAckStatus *ack_reply) {
+  TEST_ASSERT_EQUAL(SYSTEM_CAN_MESSAGE_LIGHTS, msg->msg_id);
+  count++;
+  return STATUS_CODE_OK;
+}
+
+StatusCode prv_test_cc_toggle_rx_cb_handler(const CanMessage *msg, void *context,
+                                            CanAckStatus *ack_reply) {
+  TEST_ASSERT_EQUAL(SYSTEM_CAN_MESSAGE_CRUISE_CONTROL_COMMAND, msg->msg_id);
+  count++;
+  return STATUS_CODE_OK;
+}
+
+void setup_test(void) {
+  TEST_ASSERT_OK(gpio_init());
+  interrupt_init();
+  event_queue_init();
+  gpio_it_init();
+  soft_timer_init();
+  TEST_ASSERT_OK(steering_digital_input_init());
+  TEST_ASSERT_OK(can_init(&s_can_storage, &can_settings));
+}
+
+void test_steering_digital_input_horn() {
+  TEST_ASSERT_OK(
+      can_register_rx_handler(SYSTEM_CAN_MESSAGE_HORN, prv_test_horn_rx_cb_handler, NULL));
+  GpioAddress *horn_address = test_get_address(STEERING_INPUT_HORN_EVENT);

```

Comment:
again, these can be #defined in the header so you don't have to have this function, goes for other instances in this file as well